### PR TITLE
Multi-input RNN and TimeDistributed wrapper

### PR DESCRIPTION
**Pull request by farizrahman4u**
Theano scan op allows looping through multiple sequences simultaneously, but this feature is currently hidden by Keras's abstraction, forcing users to concatenate sequences(followed by unpacking them in the step function).
- Backward compatible. The API is pretty simple; now the `inputs` argument can be a tensor or a list of tensors with same input lengths.

### CODE REVIEW

**Comment by farizrahman4u** [comment]
Also see #3057

**Comment by farizrahman4u** [comment]
ping @fchollet

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -1033,11 +1033,23 @@ def rnn(step_function, inputs, initial_states,
...
+    if type(inputs) == list:
+        input_list = []
+        for x in inputs:
+            ndim = len(x.get_shape())
+            assert ndim >= 3, "Input should be at least 3D."
+            axes = [1, 0] + list(range(2, ndim))
+            x = tf.transpose(x, (axes))
+            input_list += [tf.unpack(x)]
+        assert len(set(map(len, input_list))) == 1, "All input sequences should be of equal length."
+        input_list = [map(lambda x: x[t], input_list) for t in range(len(input_list[0]))]
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: If feel like this should be handled more elegantly in the above loop. This line is not very readable.

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -1033,11 +1033,20 @@ def rnn(step_function, inputs, initial_states,
...
-    axes = [1, 0] + list(range(2, ndim))
-    inputs = tf.transpose(inputs, (axes))
-    input_list = tf.unpack(inputs)
+    if type(inputs) == list:
+        input_list = [[]] * input_length    
+        for x in inputs:
+            ndim = len(x.get_shape())
+            assert ndim >= 3, "Input should be at least 3D."
+            for t in range(input_length):
+                    input_list[t] += [x[(slice(None), t) + (slice(None),) * (ndim - 2)]]
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Indentation issue. Are you using a linter (like pylint)?

**Comment by fchollet** [comment]
Can you describe what is the motivation for this feature? Would it be used anywhere?

**Comment by farizrahman4u** [comment]
I currently use it in a couple of custom layers. It comes at the cost of readability though.

File: keras/backend/tensorflow_backend.py [hunk 3]
```diff
@@ -1033,11 +1033,20 @@ def rnn(step_function, inputs, initial_states,
         new_states: list of tensors, latest states returned by
             the step function, of shape (samples, ...).
     '''
-    ndim = len(inputs.get_shape())
-    assert ndim >= 3, "Input should be at least 3D."
-    axes = [1, 0] + list(range(2, ndim))
-    inputs = tf.transpose(inputs, (axes))
-    input_list = tf.unpack(inputs)
+    if type(inputs) == list:
+        input_list = [[]] * input_length
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]:
```sh
>>> input_list = [[]] * 10
>>> input_list[0] += [1]
>>> input_list
[[1], [1], [1], [1], [1], [1], [1], [1], [1], [1]]
```

Probably not what you intended.

**Comment by matt-gardner** [comment]
Here's another use case for this feature: we are building reading comprehension models and memory networks using Keras.  To answer a question, we perform some multi-step operations on the input question, answer candidate, and a passage of text.  If the question is multiple choice, it would be _really_ convenient to be able to just `TimeDistribute` each of those multi-step operations, and have a softmax at the end to select an answer option.  But because `TimeDistributed` only takes a single input, we have to go through a bunch of contortions of merging the inputs in funny ways and unpacking them in our `Layer` classes.  It would simplify a whole lot of parts of our code if we could just pass several inputs into `TimeDistributed`.

**Comment by farizrahman4u** [comment]
@matt-gardner Unfortunately, this PR is way too old for me to get back at it. You can add your unpacking logic as a Lambda layer to a model, and then add your layer to the model. Since models are layers by inheritance, you can timedistribute the model.
